test/services/TxSender.test.ts

```diff
@@ -0,0 +1,157 @@
+};
+
+describe('TxSender', () => {
+  describe('when waitForBlockTime is not provided', () => {
+    it('sets 500ms as default', () => {
+      const subject = new TxSender({ wallet: {} as Wallet, chainId: 'eth' });
+      expect(subject.waitForBlockTime).to.be.equal(500);
+    });
+  });
+
+  describe('#apply', () => {
+    let fakeWallet: Wallet;
+    let fakeProvider: StubbedInstance<providers.BaseProvider>;
+    let txSender: TxSender;
+
+    before(() => {
+      fakeProvider = stubConstructor(providers.JsonRpcProvider);
+      fakeWallet = Wallet.createRandom().connect(fakeProvider);
+    });
+
+    beforeEach(() => {
+      // TxSender needs a new block number every time
+      for (let i = 0; i < 10; i++) {
+        fakeProvider.getBlockNumber.onCall(i /* calls start at 0 */).resolves(100_000 + i);
+      }
+
+      sinon.stub(GasEstimator, 'apply').resolves({
+        gasPrice: 100,
+        maxPriorityFeePerGas: 100,
+        maxFeePerGas: 100,
+        isTxType2: true
+      } as GasEstimation);
+    });
+
+    afterEach(() => {
+      fakeProvider.getBlockNumber.reset();
+      sinon.restore();
+    });
+
+    describe('when transaction is submitted successfully', () => {
+      before(() => {
+        txSender = new TxSender({ wallet: fakeWallet, chainId: 'eth' });
+      });
+
+      //* This test is not exiting. This suggests
+      //* that one asynchronous operation is not being finalized.
+      it('resolves the transaction receipt', async () => {
+        const fakeTransactionFunction = makeFakeTransactionFunction({ status: 1 });
+        const subject = await txSender.apply(fakeTransactionFunction, 100, 500, .01);
+        expect(subject.status).to.be.equal(1);
+      });
+    });
+
+    describe('when transaction is not submitted successfully', () => {
+      beforeEach(() => {
+        // casting is necessary to ensure compatibility.
+        fakeWallet = new FakeWallet(fakeProvider) as unknown as Wallet;
+        txSender = new TxSender({ wallet: fakeWallet, chainId: 'eth' });
+      });
+
+      it('attempts to cancel the pending transaction', async () => {
+        const spy = sinon.spy(fakeWallet, 'getTransactionCount');
+        const fakeTransactionFunction = makeFakeTransactionFunction();
+
+        try {
+          await txSender.apply(fakeTransactionFunction, 100, 500, .01);
+        } catch {}
+
+        expect(spy.called).to.be.true;
+      });
+
+      it('throws a timeout error', async () => {
+        const fakeTransactionFunction = makeFakeTransactionFunction();
+
+        const subject = txSender.apply(fakeTransactionFunction, 100, 500, .01);
+        expect(subject).to.eventually.throw('mint TX timeout');
+      });
+
+      describe('when transaction is canceled successfully', () => {
+        it('does not try to re-send the cancelling transaction', async () => {
+          const expectedCallCount = 4;
+          const fakeTransactionFunction = makeFakeTransactionFunction();
+
+          try {
```

Comment:
I think you can remove the try-catch.

Reply:
I can't. When the transaction is cancelled, it throws an error anyway.

Reply:
Removing it will break both tests. (resolved the other comment)